lucene/core/src/java/org/apache/lucene/search/LRUQueryCache.java

```diff
@@ -732,8 +737,38 @@ public ScorerSupplier scorerSupplier(LeafReaderContext context) throws IOExcepti
 
       if (docIdSet == null) {
         if (policy.shouldCache(in.getQuery())) {
-          docIdSet = cache(context);
-          putIfAbsent(in.getQuery(), docIdSet, cacheHelper);
+          final ScorerSupplier supplier = in.scorerSupplier(context);
+          if (supplier == null) {
+            putIfAbsent(in.getQuery(), DocIdSet.EMPTY, cacheHelper);
+            return null;
+          }
+
+          final long cost = supplier.cost();
+          return new ScorerSupplier() {
+            @Override
+            public Scorer get(long leadCost) throws IOException {
+              // skip cache operation which would slow query down too much
+              if (cost > leadCost * skipCacheFactor) {
```

Comment:
Reading your tests made me wonder that the multiplication would overflow here when `Long.MAX_VALUE` is used for `skipCacheFactor`.

Reply:
Yes, the multiplication may overflow. Considering _leadCost_ may be zero, I will use `cost / skipCacheFactor > leadCost` to replace the multiplication.

BTW, I found out that why my tests success is that _leadCost_ was always 1. I will strengthen my tests a little.